README.md
### WURFL Logstash plugin

This project contains a java plugin for Logstash that enriches a stream of data with device data obtained from WURFL ([through WURFL Microservice Client, as deployed in one of multiple ways](https://www.scientiamobile.com/products/wurfl-microservice/)).

This plugin requires Java 8+, Gradle 5.x or above and has been tested with Logstash 8.0.0 and 7.6.x
It also requires a running instance of WURFL Microservice server, available on [AWS](https://aws.amazon.com/marketplace/search/results?x=26&y=17&searchTerms=WURFL), [Azure](https://azuremarketplace.microsoft.com/en-us/marketplace/apps?search=WURFL&page=1) or as a [Docker container](https://www.scientiamobile.com/products/wurfl-microservice-docker-detect-device/)

## Compile the project

From the root of the project enter ` ./gradlew gem`

The compiler will create a file `logstash-filter-logstash_filter_wurfl_device_detection-x.y.z.gem`, where x.y.z version number
is the one defined in the `VERSION` file.

## Install the plugin on logstash

  From the logstash installation `bin` directory execute
  `./logstash-plugin install --local <plugin_project_home>/logstash-filter-logstash_filter_wurfl_device_detection-x.y.z.gem`
  
  You can find a sample configuration for this plugin under `sample_config/wurfl_filter_config_with_http_input`; you will want to create your own
  production configuration file using input and output plugins of your choice.
  
### Example configuration - using HTTP input plugin to receive HTTP request data
Scenario: we configure logstash to receive HTTP request information that we want to enrich with WURFL data.

```
input {
  http {
    host => "0.0.0.0"
    port => "19080"
  }
}
filter {
  logstash_filter_wurfl_device_detection {
    source => "headers"
    cache_size => 300000
    inject_wurfl_id => true
    inject_wurfl_info => false
    inject_wurfl_api_version => false
    scheme => "http"
    host => "localhost"
    port => "8080"
  }
}
output {
  stdout { codec => rubydebug }
}
```

The http input plugin receives an http request to the specified host and port, with a payload map that contains the HTTP headers that the WURFL plugin will analyze.
Note that the `source` name is `headers`. Also note that you can configure the logstash input as you want,
but if you want the WURFL plugin to work with headers, you must configure it so that it uses an header map.

### Example configuration - using JSON input and filter plugins to receive HTTP request data
Scenario: we configure logstash to receive HTTP request information from one or more JSON files (needs logstash-filter-json plugin).

```
input {
  file {
    type => "json"
    path => "<path to>/logstash-filter-wurfl/sample_input/kafka_events_mini.json"
    start_position => "beginning"
  }
}
filter {
    json {
    source => "message"
}
}
filter {
  wurfl_device_detection {
    source => "message"
    cache_size => 300000
    inject_wurfl_id => true
    inject_wurfl_info => false
    inject_wurfl_api_version => false
    # commenting the capabilities config, you get them all
    #static_capabilities => ["model_name", "brand_name"]
    #virtual_capabilities => ["form_factor"]
    scheme => "http"
    host => "localhost"
    port => "8080"
  }
}
output {
  # you may choose whatever output you want
  stdout { codec => rubydebug }
}
```

Source field "message" is the root element of the json file

### Starting logstash

Start the WURFL Microservice server on AWS/Azure/Docker then, let's run logstash:

`./logstash -f <path_to_configuration>.conf>`

in case you choose the HTTP input configuration, sending a HTTP request to its configured port (in this case 19080) like this (we just add the user-agent header for simplicity, but you can send all headers using the -H flag):

`curl -H "User-Agent: Mozilla/5.0 (Linux; Android 8.0.0; Pixel 2 XL Build/OPD1.170816.004) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3359.139 Mobile Safari/537.36" http://localhost:19080`

we'll get an output that looks like this:
  
  ```
{
       "message" => "",
      "@version" => "1",
       "headers" => {
         "request_method" => "GET",
              "http_host" => "localhost:19080",
            "http_accept" => "*/*",
        "http_user_agent" => "Mozilla/5.0 (Linux; Android 8.0.0; Pixel 2 XL Build/OPD1.170816.004) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3359.139 Mobile Safari/537.36",
         "content_length" => "0",
           "request_path" => "/",
           "http_version" => "HTTP/1.1"
    },
    "@timestamp" => 2020-03-19T09:01:58.761Z,
          "host" => "127.0.0.1",
         "wurfl" => {
                            "is_robot" => "false",
                              "is_ios" => "false",
                  "advertised_browser" => "Chrome Mobile",
                         "form_factor" => "Smartphone",
                      "is_app_webview" => "false",
                           "is_mobile" => "true",
                          "is_android" => "true",
                          "brand_name" => "Google",
                   "resolution_height" => "2560",
                    "is_windows_phone" => "false",
                            "wurfl_id" => "google_pixel_2_xl_ver1",
                              "is_app" => "false",
                "complete_device_name" => "Google Pixel 2 XL",
                     "pointing_method" => "touchscreen",
                           "device_os" => "Android",
                 "advertised_app_name" => "Chrome browser",
                         "device_name" => "Google Pixel 2 XL",
                    "resolution_width" => "1440",
        "advertised_device_os_version" => "8.0.0",
                           "is_tablet" => "false",
                          "model_name" => "Pixel 2 XL",
                      "mobile_browser" => "Chrome Mobile",
                          "is_smarttv" => "false",
                      "is_touchscreen" => "true",
          "advertised_browser_version" => "65.0.3359.139",
                       "is_smartphone" => "true",
                "advertised_device_os" => "Android",
        [...]
    }
}

```

Logstash will automatically parse the JSON input file configuration and print out the following.

- `stdin` and `stdout` define which input and output plugin will be used:  in our scenario
 we use the HTTP input plugin, while we use the ruby debug console as output.
- `cache_size` (integer) is the size of the WURFL Microservice client cache. Defaults to 100000
- `inject_wurfl_id` defines whether `wurfl_id` will be added to enriched output (defaults to true)
- `inject_wurfl_info` defines whether `wurfl_info` will be added to enriched output (defaults to false)
- `inject_wurfl_api_version` defines whether `wurfl_api_version` will be added to enriched output (defaults to false)
- `static_capabilities` defines the list of static capabilities that you want to add to the enriched output (defaults to all)
- `virtual_capabilities` defines the list of virtual capabilities that you want to add to the enriched output (defaults to all)
- `scheme` defines the connection scheme to use to connect to WURFL Microservice server (currently only HTTP is supported)
- `host` host/ip address of the WURFL Microservice server (defaults to localhost)
- `port` port of the WURFL Microservice server (defaults to 80)


CHANGELOG.md
- 0.4.0 - now support for array of events

- 0.3.4 - added gemspec file, fixed naming convention issues

- 0.3.3 - Better description of WM pre requisites in rubygem.org deploy

- 0.3.2 - Fixed wrong info in build file

- 0.3.1 - Small logging improvements

- 0.3.0 - Initial release
 


src/test/java/com/scientiamobile/logstash/WurflDeviceDetectionTest.java
package com.scientiamobile.logstash;

import co.elastic.logstash.api.Configuration;
import co.elastic.logstash.api.Event;
import org.jruby.RubyString;
import org.junit.BeforeClass;
import org.junit.Test;
import org.logstash.plugins.ConfigurationImpl;

import java.util.*;

import static com.scientiamobile.logstash.Config.TARGET_CONFIG;
import static org.junit.Assert.*;

/*
 NOTE: this test class needs to connect to a running WURFL Microservice server
 */
public class WurflDeviceDetectionTest {

    private static String testHost = "localhost";
    private static String testPort = "8080";

    @BeforeClass
    public static void init(){
        String h = System.getenv("TEST_HOST");
        String p = System.getenv("TEST_PORT");
        if(h!= null && h.length() >0){
            testHost = h;
        }

        if(p!= null && p.length() >0){
            testPort = p;
        }
        System.out.println("TEST HOST: " + testHost);
        System.out.println("TEST PORT: " + testPort);
    }

    @Test
    public void filterTestWithFullConfig(){

        List<String> staticCaps = new ArrayList<>();
        staticCaps.add("brand_name");
        staticCaps.add("model_name");

        List<String> vCaps = new ArrayList<>();
        vCaps.add("form_factor");

        Map<String, Object> testConfig = new HashMap<>();
        testConfig.put("cache_size", new Long(10000));
        testConfig.put("source", "headers");
        testConfig.put("inject_wurfl_id", Boolean.TRUE);
        testConfig.put("inject_wurfl_info", Boolean.TRUE);
        testConfig.put("inject_wurfl_api_version", Boolean.TRUE);
        testConfig.put("static_capabilities", staticCaps);
        testConfig.put("virtual_capabilities", vCaps);
        testConfig.put("scheme", "http");
        testConfig.put("host", testHost);
        testConfig.put("port", testPort);

        Configuration config = new ConfigurationImpl(testConfig);
        WurflDeviceDetection filter = new WurflDeviceDetection("test_id", config, null);

        // Create mock event
        Event e = new org.logstash.Event();
        Map<String, String> headers = new HashMap<>();
        headers.put("User-Agent", "Mozilla/5.0 (Linux; Android 5.1; ASUS_Z00VD Build/LMY47I; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/47.0.2526.100 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/108.0.0.17.68;]");
        e.setField("headers", headers);
        Collection<Event> events = filter.filter(Collections.singletonList(e), null);
        Collection<Event> enrichedEvents = filter.filter(events, null);
        assertNotNull(enrichedEvents);
        assertEquals(1, enrichedEvents.size());
        enrichedEvents.forEach(event -> {
            Map<String, Object> map = event.getData();
            assertNotNull(map);
            Object target = map.get(TARGET_CONFIG);
            assertNotNull(target);
            Map<String, RubyString> targetMap = (Map<String, RubyString>)target;
            assertEquals("asus_z00vd_ver1", targetMap.get("wurfl_id").asJavaString());
            assertEquals("Asus", targetMap.get("brand_name").asJavaString());
            assertEquals("Z00VD", targetMap.get("model_name").asJavaString());
            assertEquals("Smartphone", targetMap.get("form_factor").asJavaString());
            assertEquals("asus_z00vd_ver1", targetMap.get("wurfl_id").asJavaString());
            assertNotNull(targetMap.get("wurfl_api_version"));
            assertNotNull(targetMap.get("wurfl_info"));

        });
    }

    @Test
    public void filterTestWithNoCapsInConfig(){

        Map<String, Object> testConfig = new HashMap<>();
        testConfig.put("cache_size", new Long(10000));
        testConfig.put("source", "headers");
        testConfig.put("inject_wurfl_id", Boolean.TRUE);
        testConfig.put("inject_wurfl_info", Boolean.TRUE);
        testConfig.put("inject_wurfl_api_version", Boolean.TRUE);
        testConfig.put("scheme", "http");
        testConfig.put("host", testHost);
        testConfig.put("port", testPort);

        Configuration config = new ConfigurationImpl(testConfig);
        WurflDeviceDetection filter = new WurflDeviceDetection("test_id", config, null);

        // Create mock event
        Event e = new org.logstash.Event();
        Map<String, String> headers = new HashMap<>();
        headers.put("User-Agent", "Mozilla/5.0 (Linux; Android 5.1; ASUS_Z00VD Build/LMY47I; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/47.0.2526.100 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/108.0.0.17.68;]");
        e.setField("headers", headers);
        Collection<Event> events = filter.filter(Collections.singletonList(e), null);
        Collection<Event> enrichedEvents = filter.filter(events, null);
        assertNotNull(enrichedEvents);
        assertEquals(1, enrichedEvents.size());
        enrichedEvents.forEach(event -> {
            Map<String, Object> map = event.getData();
            assertNotNull(map);
            Object target = map.get(TARGET_CONFIG);
            assertNotNull(target);
            Map<String, RubyString> targetMap = (Map<String, RubyString>)target;
            assertNotNull(targetMap.get("wurfl_api_version"));
            assertNotNull(targetMap.get("wurfl_info"));
            // When a list of capabilities is NOT specified, all of them are returned
            assertTrue(targetMap.size() >= 40);
        });
    }

    @Test
    public void filterTestWithHttpHeadersConfig(){

        Map<String, Object> testConfig = new HashMap<>();
        testConfig.put("cache_size", new Long(10000));
        testConfig.put("source", "headers");
        testConfig.put("inject_wurfl_id", Boolean.TRUE);
        testConfig.put("inject_wurfl_info", Boolean.TRUE);
        testConfig.put("inject_wurfl_api_version", Boolean.TRUE);
        testConfig.put("scheme", "http");
        testConfig.put("host", testHost);
        testConfig.put("port", testPort);

        Configuration config = new ConfigurationImpl(testConfig);
        WurflDeviceDetection filter = new WurflDeviceDetection("test_id", config, null);

        // Create mock event
        Event e = new org.logstash.Event();
        Map<String,String> headers = new HashMap<>();
        headers.put("http_User_Agent".toLowerCase(), "Mozilla/5.0 (Linux; Android 5.1; ASUS_Z00VD Build/LMY47I; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/47.0.2526.100 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/108.0.0.17.68;]");
        headers.put("X_OperaMini_Phone_UA".toLowerCase(), "test_x_header");
        e.setField("headers", headers);
        Collection<Event> events = filter.filter(Collections.singletonList(e), null);
        Collection<Event> enrichedEvents = filter.filter(events, null);
        assertNotNull(enrichedEvents);
        assertEquals(1, enrichedEvents.size());
        enrichedEvents.forEach(event -> {
            Map<String, Object> map = event.getData();
            assertNotNull(map);
            Object target = map.get(TARGET_CONFIG);
            assertNotNull(target);
            Map<String, RubyString> targetMap = (Map<String, RubyString>)target;
            assertNotNull(targetMap.get("wurfl_api_version"));
            assertNotNull(targetMap.get("wurfl_info"));
            // When a list of capabilities is NOT specified, all of them are returned
            assertTrue(targetMap.size() >= 40);
        });
    }

    @Test
    public void filterTestWithNoWurflInfoInConfig(){

        // This test does not add WURFL ID nor WURFL Info and version strings
        Map<String, Object> testConfig = new HashMap<>();
        testConfig.put("cache_size", new Long(10000));
        testConfig.put("source", "headers");
        testConfig.put("scheme", "http");
        testConfig.put("host", testHost);
        testConfig.put("port", testPort);
        testConfig.put("inject_wurfl_id", "false"); // we need to specify this because default is true

        Configuration config = new ConfigurationImpl(testConfig);
        WurflDeviceDetection filter = new WurflDeviceDetection("test_id", config, null);

        // Create mock event
        Event e = new org.logstash.Event();
        Map<String, String> headers = new HashMap<>();
        headers.put("User-Agent", "Mozilla/5.0 (Linux; Android 5.1; ASUS_Z00VD Build/LMY47I; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/47.0.2526.100 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/108.0.0.17.68;]");
        e.setField("headers", headers);
        Collection<Event> events = filter.filter(Collections.singletonList(e), null);
        Collection<Event> enrichedEvents = filter.filter(events, null);
        assertNotNull(enrichedEvents);
        assertEquals(1, enrichedEvents.size());
        enrichedEvents.forEach(event -> {
            Map<String, Object> map = event.getData();
            assertNotNull(map);
            Object target = map.get(TARGET_CONFIG);
            assertNotNull(target);
            Map<String, RubyString> targetMap = (Map<String, RubyString>)target;
            assertNull(targetMap.get("wurfl_api_version"));
            assertNull(targetMap.get("wurfl_info"));
            assertNull(targetMap.get("wurfl_id"));
            // When a list of capabilities is NOT specified, all of them are returned
            assertTrue(targetMap.size() >= 40);
        });
    }

    @Test(expected = IllegalArgumentException.class)
    public void filterTestWithNoServerConfig(){

        Map<String, Object> testConfig = new HashMap<>();
        testConfig.put("cache_size", new Long(10000));
        testConfig.put("source", "headers");

        Configuration config = new ConfigurationImpl(testConfig);
        new WurflDeviceDetection("test_id", config, null);
    }

    @Test
    public void filterTestWithoutSourceConfig(){

        Map<String, Object> testConfig = new HashMap<>();
        testConfig.put("scheme", "http");
        testConfig.put("host", testHost);
        testConfig.put("port", testPort);

        Configuration config = new ConfigurationImpl(testConfig);
        WurflDeviceDetection filter = new WurflDeviceDetection("test_id", config, null);

        // Create mock event
        Event e = new org.logstash.Event();
        Map<String, String> headers = new HashMap<>();
        headers.put("User-Agent", "Mozilla/5.0 (Linux; Android 5.1; ASUS_Z00VD Build/LMY47I; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/47.0.2526.100 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/108.0.0.17.68;]");
        e.setField("headers", headers);
        Collection<Event> events = filter.filter(Collections.singletonList(e), null);
        Collection<Event> enrichedEvents = filter.filter(events, null);
        assertNotNull(enrichedEvents);
        assertEquals(1, enrichedEvents.size());
        enrichedEvents.forEach(event -> {
            Map<String, Object> map = event.getData();
            assertNotNull(map);
            Object target = map.get(TARGET_CONFIG);
            assertNotNull(target);
            Map<String, RubyString> targetMap = (Map<String, RubyString>)target;
            assertNull(targetMap.get("wurfl_api_version"));
            assertNull(targetMap.get("wurfl_info"));
            // When a list of capabilities is NOT specified, all of them are returned
            assertTrue(targetMap.size() >= 40);
        });

        // Case two, an event with a different source name ("agent")
        e = new org.logstash.Event();
        // We put the data in "headers" source, thus test will pass because "headers" is the default name for source
        e.setField("agent", "Mozilla/5.0 (Linux; Android 5.1; ASUS_Z00VD Build/LMY47I; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/47.0.2526.100 Mobile Safari/537.36 [FB_IAB/FB4A;FBAV/108.0.0.17.68;]");
        events = filter.filter(Collections.singletonList(e), null);
        enrichedEvents = filter.filter(events, null);
        enrichedEvents.forEach(event -> {
            Map<String, Object> map = event.getData();
            assertNotNull(map);
            Object target = map.get(TARGET_CONFIG);
            // In this case, since "headers" has not been found, the event won't be enriched with WURFL data, which results in an empty "wurfl" data element
            assertNull(target);

        });
    }

    @Test
    public void filterTestWithWrongCacheConfig(){

        Map<String, Object> testConfig = new HashMap<>();
        testConfig.put("cache_size", new Long(-1));
        testConfig.put("source", "headers");

        Configuration c = new ConfigurationImpl(testConfig);
        // Wrong cache size value, default value is used
        Config config = new Config(c);
        assertEquals(Config.DEFAULT_CACHE_SIZE.intValue(), config.getCacheSize().intValue());
    }
}


src/main/java/com/scientiamobile/logstash/WurflDeviceDetection.java
package com.scientiamobile.logstash;

import co.elastic.logstash.api.*;
import com.scientiamobile.wurfl.wmclient.*;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.servlet.http.HttpServletRequest;
import java.util.*;

import static com.scientiamobile.logstash.Config.*;

// Filter name uses underscores because classname and plugin names MUST match with the exception of casing and underscores
@LogstashPlugin(name = "wurfl_device_detection")
public class WurflDeviceDetection implements Filter {

    static final String WURFL_ID = "wurfl_id";
    static final String WURFL_INFO = "wurfl_info";
    static final String TAG_PARSE_FAILURE = "_wurflparsefailure";
    static final String WURFL_API_VERSION = "wurfl_api_version";
    private static Logger logger = LogManager.getLogger(WurflDeviceDetection.class);


    private final Config configuration;
    private String id;
    private WmClient wmClient;

    public WurflDeviceDetection(String id, Configuration lconfig, Context ctx) {
        this(id, lconfig);
    }

    public WurflDeviceDetection(String id, Configuration lconfig) {
        this.id = id;
        this.configuration = new Config(lconfig);
        initWmClient(configuration);
        logger.debug("WURFL microservice client created");
    }

    /*
     Initializes the WURFL Microservice client (server to which it connects, cache size, capabilities to be returned).
     */
    private void initWmClient(Config config) {
        try {
            wmClient = WmClient.create(config.getScheme(), config.getHost(), config.getPort(), "");
            wmClient.setCacheSize(config.getCacheSize());
            if (ArrayUtils.isNotEmpty(config.getStaticCapabilities())) {
                wmClient.setRequestedStaticCapabilities(config.getStaticCapabilities());
            }
            if (ArrayUtils.isNotEmpty(config.getVirtualCapabilities())) {
                wmClient.setRequestedVirtualCapabilities(config.getVirtualCapabilities());
            }
        } catch (Exception e) {
            throw new IllegalArgumentException("[WURFL filter]: An error occurred initializing WURFL microservice client, please check your client configuration "
                    + e.getMessage(), e);
        }
    }

    /**
     * Extracts the HTTP request headers and uses them to query the WURFL Microservice server to get device detection information, which are added
     * to the event output data.
     *
     * @param collection          a collection of events
     * @param filterMatchListener this is unused
     * @return the collection of events, enriched with device detection data
     */
    @Override
    public Collection<Event> filter(Collection<Event> collection, FilterMatchListener filterMatchListener) {

        collection.forEach(event -> {
            Model.JSONInfoData info;
            try {
                info = wmClient.getInfo();
                Object o = event.getField(configuration.getSource());
                // This happens if source field has been wrongly configured (ie: source is "headers", but event data are in a field with another name
                if (o == null) {
                    logger.error("Message object taken from source  [" + configuration.getSource() + "] is null, check your plugin configuration");
                    event.tag(TAG_PARSE_FAILURE);
                    return;
                }

                Model.JSONDeviceData tdevice = null;
                final List<Model.JSONDeviceData> tdevices = new ArrayList<>();

                // We usually get a Map when a single event is sent
                if (o instanceof Map) {
                    HttpServletRequest req = new HttpServletRequestWrapper((Map) o);
                    tdevice = wmClient.lookupRequest(req);
                    final Model.JSONDeviceData device = tdevice;
                    // Device can be null if WM server version is < 2.1.0
                    if (device == null) {
                        event.tag(TAG_PARSE_FAILURE);
                        return;
                    }
                    Map<String, String> data = fillDataMap(info, device);
                    event.setField(TARGET_CONFIG, data);
                    // We usually get an ArrayList when we get an array or set of events
                } else if (o instanceof ArrayList) {
                    final List<Map<String, String>> dataMaps = new ArrayList<>();
                    ArrayList<Map> maps = (ArrayList<Map>) o;
                    for (Map m : maps) {
                        HttpServletRequest req = new HttpServletRequestWrapper((Map) m.get("headers"));
                        tdevice = wmClient.lookupRequest(req);
                        if (tdevice == null) {
                            continue;
                        }
                        tdevices.add(tdevice);
                        dataMaps.add(fillDataMap(info, tdevice));
                    }
                    event.setField(TARGET_CONFIG, dataMaps);
                } else {
                    logger.error("Input object type: " + o.getClass().getCanonicalName());
                    logger.error("Message source name  [" + configuration.getSource() + "] is not of type ArrayList or Map, check your plugin configuration");
                }
            } catch (WmException e) {
                // Just log and go on
                logger.error("dropping event " + event.toString() + " due to " + e.getLocalizedMessage(), e);
            }
        });

        return collection;
    }

    private Map<String, String> fillDataMap(Model.JSONInfoData info, Model.JSONDeviceData device) {
        // Device has been detected, let's collect device info and capabilities
        Map<String, String> data = new HashMap<>();
        if (configuration.getInjectWurflId()) {
            data.put(WURFL_ID, device.capabilities.get(WURFL_ID));
        }

        if (configuration.getInjectWurflInfo()) {
            data.put(WURFL_INFO, info.getWurflInfo());
        }

        if (configuration.getInjectWurflApiVersion()) {
            data.put(WURFL_API_VERSION, info.getWurflApiVersion());
        }

        // Add capabilities to data
        if (configuration.getInjectWurflId()) {
            device.capabilities.keySet().forEach(name -> {
                data.put(name, device.capabilities.get(name));
            });
        } else { // inject_wurfl_id not enabled, we filter it from capability list
            device.capabilities.keySet().stream().filter(name -> !"wurfl_id".equals(name)).forEach(name -> {
                data.put(name, device.capabilities.get(name));
            });
        }
        return data;
    }

    @Override
    public Collection<PluginConfigSpec<?>> configSchema() {
        List<PluginConfigSpec<?>> configs = new ArrayList<>();
        configs.add(SOURCE_CONFIG);
        configs.add(CACHE_SIZE_CONFIG);
        configs.add(WM_CONN_HOST_CONFIG);
        configs.add(WM_CONN_SCHEME_CONFIG);
        configs.add(WM_CONN_PORT_CONFIG);
        configs.add(INJECT_WURFL_API_VERSION);
        configs.add(INJECT_WURFL_ID);
        configs.add(INJECT_WURFL_INFO);
        configs.add(STATIC_CAPABILITIES_CONFIG);
        configs.add(VIRTUAL_CAPABILITIES_CONFIG);

        return configs;
    }

    @Override
    public String getId() {
        return this.id;
    }
}


src/main/java/com/scientiamobile/logstash/Config.java
package com.scientiamobile.logstash;

import co.elastic.logstash.api.Configuration;
import co.elastic.logstash.api.PluginConfigSpec;
import java.util.ArrayList;
import java.util.List;

/**
 * Holds configuration information for WURFL device detection Logstash plugin
 */
class Config {

    private static final List<Object> DEFAULT_VIRTUAL_CAPABILITIES = new ArrayList<>();
    private static List<Object> DEFAULT_STATIC_CAPABILITIES = new ArrayList<>();
    static final Integer DEFAULT_CACHE_SIZE = 100000;

    static final PluginConfigSpec<String> SOURCE_CONFIG =
            PluginConfigSpec.stringSetting("source", "headers");
    static final PluginConfigSpec<Long> CACHE_SIZE_CONFIG =
            PluginConfigSpec.numSetting("cache_size", DEFAULT_CACHE_SIZE);
    static final PluginConfigSpec<String> WM_CONN_SCHEME_CONFIG =
            PluginConfigSpec.stringSetting("scheme", "http");
    static final PluginConfigSpec<String> WM_CONN_HOST_CONFIG =
            PluginConfigSpec.stringSetting("host", "localhost");
    static final PluginConfigSpec<String> WM_CONN_PORT_CONFIG =
            PluginConfigSpec.stringSetting("port", "80");
    static final PluginConfigSpec<Boolean> INJECT_WURFL_ID =
            PluginConfigSpec.booleanSetting("inject_wurfl_id", true);
    static final PluginConfigSpec<Boolean> INJECT_WURFL_INFO =
            PluginConfigSpec.booleanSetting("inject_wurfl_info", false);
    static final PluginConfigSpec<Boolean> INJECT_WURFL_API_VERSION =
            PluginConfigSpec.booleanSetting("inject_wurfl_api_version", false);

    static final PluginConfigSpec<List<Object>> STATIC_CAPABILITIES_CONFIG =
            PluginConfigSpec.arraySetting("static_capabilities", DEFAULT_STATIC_CAPABILITIES, false, false);
    static final PluginConfigSpec<List<Object>> VIRTUAL_CAPABILITIES_CONFIG =
            PluginConfigSpec.arraySetting("virtual_capabilities", DEFAULT_VIRTUAL_CAPABILITIES, false, false);

    static final String TARGET_CONFIG = "wurfl";

    private String   source;
    private Integer  cacheSize;
    private Boolean  injectWurflId;
    private Boolean  injectWurflInfo;
    private Boolean  injectWurflApiVersion;
    private String[] staticCapabilities;
    private String[] virtualCapabilities;
    private String   scheme;
    private String   host;
    private String   port;

    public Config(Configuration lconfig){

        source = lconfig.get(SOURCE_CONFIG);

        // WM client connection info
        scheme = lconfig.get(WM_CONN_SCHEME_CONFIG);
        host   = lconfig.get(WM_CONN_HOST_CONFIG);
        port = lconfig.get(WM_CONN_PORT_CONFIG);

        // WM client Cache configuration
        cacheSize = lconfig.get(CACHE_SIZE_CONFIG).intValue();
        if (cacheSize <= 0) {
            cacheSize = DEFAULT_CACHE_SIZE;
        }

        // WM data inject configuration
        injectWurflId = lconfig.get(INJECT_WURFL_ID);
        injectWurflInfo = lconfig.get(INJECT_WURFL_INFO);
        injectWurflApiVersion = lconfig.get(INJECT_WURFL_API_VERSION);

        // WM capabilities configuration
        staticCapabilities = lconfig.get(STATIC_CAPABILITIES_CONFIG).toArray(new String[0]);
        virtualCapabilities = lconfig.get(VIRTUAL_CAPABILITIES_CONFIG).toArray(new String[0]);

    }

    public String getSource() {
        return source;
    }

    public Integer getCacheSize() {
        return cacheSize;
    }

    public Boolean getInjectWurflId() {
        return injectWurflId;
    }

    public Boolean getInjectWurflInfo() {
        return injectWurflInfo;
    }

    public Boolean getInjectWurflApiVersion() {
        return injectWurflApiVersion;
    }

    public String[] getStaticCapabilities() {
        return staticCapabilities;
    }

    public String[] getVirtualCapabilities() {
        return virtualCapabilities;
    }

    public String getScheme() {
        return scheme;
    }

    public String getHost() {
        return host;
    }

    public String getPort() {
        return port;
    }
}